### ISSUE TITLE
Add option to exclude variables from export

### ISSUE THREAD

**Issue by jamesrkg** [Crunch-io/scrunch#93]
Currently we can give a list of variables to *include* using `ds.export(..., variables)` but recent experience needing to exclude a handful of variables suggests adding the ability to give a list to *exclude* instead (and get everything else) would save a lot of time in those situations.

**Comment by jamesrkg** [comment]
Since you wouldn't need to include and exclude at the same time, a bool to control the interpretation of `variables` may be a good approach, set to include by default.

**Comment by xbito** [comment]
Jj is telling me that there is a "deselect" function in the API that would be ideal in this case. That way we don't need to derive the inclusion list from the exclusion one.

I'm not against the boolean approach, but it doesn't make me happy either. Lets see what the assigned person to the ticket proposes later.

**Comment by jamesrkg** [comment]
@xbito however we get there is fine by me. 👍

**Comment by xbito** [comment]
The export method looks like this:
def export(self, path, format='csv', filter=None, variables=None,
               hidden=False, options=None, metadata_path=None):

It is turning into too many parameters, some of those can't even be sent together, like a list of variables with a hidden=True, it doesn't make sense. The method itself is starting to have too many paths, too many if conditions.

But on the other end, I don't have a very good proposal on how to refactor this out, maybe an Export class would make it a bit cleaner?

I'll ask @jjdelc for his opinion.

**Comment by jjdelc** [comment]
Maybe we should split into two things, a helper that maps out to scrunch the dataset export endpoint and handles the file downloading and wait (probably similar to pycrunch.exporting module) and another helper to only handle the complexity of variable selection.

The API only selects a `where` expression for variable selection which translates in a `select` or `deselect` function. Right now that expression is being built inside the export function and the combination of flags is becoming a problem.

We could have a helper or class that generates the needed `where` expression.

What are the use cases? 

 * Selection of individual variables
 * Everything including hidden
 * Everything without hidden
 * Everything except a selection

**Comment by jamesrkg** [comment]
@jjdelc in terms of the use cases, if you're excluding some variables you would also want the ability to export (or not) all the hidden variables you're _not_ excluding, so I'd change it slightly to:

* Selection of individual variables
* Everything including hidden
* Everything without hidden
* Everything except a selection, including hidden
* Everything except a selection, without hidden

@xbito i agree it's starting to look a bit messy, but I haven't managed to think of an alternative that is less complicated. 

That said, the complications aren't too extensive:

* If specific variables are included then hidden is moot
* Giving only one of including or excluding variables make sense

This doesn't seem completely horrible when the alternative may be to force the user to manage a new class or multiple methods, either of which just shifts this complication into the hands of the user.

If we accept that managing these conflicts internally is the lesser evil then replacing `variables` with `select` and adding `deselect` may be a reasonable solution..

Based on the use cases above that translates to:

* Selection of individual variables
```python
ds.export(select=varlist)
```

* Everything including hidden
```python
ds.export(hidden=True)
```

* Everything without hidden
```python
ds.export(hidden=False)
```

* Everything except a selection, including hidden
```python
ds.export(deselect=varlist, hidden=True)
```

* Everything except a selection, without hidden
```python
ds.export(deselect=varlist, hidden=False)
```

From a scrunch-user's pov the above is simple and straighforward even though internally the design may seem somewhat conflicted.

**Comment by xbito** [comment]
Those examples don't look very heavy, if only they were the actual use cases, we have to remember the current signature of the function:

```
def export(self, path, format='csv', filter=None, variables=None,
hidden=False, options=None, metadata_path=None):
```

I'm inclined to:

- remove the metadata export, do it in a separate helper.
- infer the format from the path
- Add the select/deselect parameters, deprecate the variables parameter.

Do you agree Jj?

**Comment by jamesrkg** [comment]
@xbito yep, my examples were simplified to highlight the specific use cases around select/deselect/hidden, the remainder is assumed.

Inferring the format from the path is a good idea. If you want to remove the metadata path that would also be fine, you could just expose it from `ds.metadata` (pointing at `.resource.table['metadata']`) and leave the user to save it themselves.

**Comment by jjdelc** [comment]
That's a very clean proposal @jamesrkg 

@xbito looks good, we can prepare the payload inside the method and optimize that to have a cleaner payload, it may be possible to translate these queries without having to go through the list of all variables. 


Example for selection and excluding hidden:
```
where={
  function: deselect,
  args: [{map: {...map of ids...}}],
  frame: {
    function: find,
    args: [{references.discarded: {not: True}}]
  }
}
```
